### PR TITLE
GOBBLIN-1107: Lazily initialize Helix TaskStateModelFactory in Gobbli…

### PR DESCRIPTION
**Pull request by sv2000**
…nTaskRunner

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1107

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
Current behavior eagerly initializes the taskStateModelFactory inside the constructor. This causes Helix task assignment to be stuck when a Helix manager connection needs to be retried.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Existing unit tests.

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by sv2000** [comment]
@autumnust please review.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinTaskRunner.java [hunk 1]
```diff
@@ -191,6 +179,17 @@ public GobblinTaskRunner(String applicationName,
...
+        this.isTaskDriver? "taskDriver" : "worker",
+        applicationName,
+        helixInstanceName,
+        applicationId,
+        taskRunnerId,
+        config,
+        appWorkDirOptional);
+  }
+
+  private synchronized TaskRunnerSuiteBase initTaskStateModelFactory() throws ReflectiveOperationException {
```
Comment (autumnust) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinTaskRunner.java:192]: this method has both return value and side effect which should be [avoided](https://softwareengineering.stackexchange.com/questions/289948/origin-of-a-method-should-return-a-value-or-have-side-effects-but-not-both). Shall we consider divide it into two: get TaskRunnerSuite  and initTaskStateModelFactory(TaskRunnerSuiteBase) if you want to make the init as a method ?

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/GobblinTaskRunnerTest.java [hunk 2]
```diff
@@ -128,21 +141,31 @@ public void testBuildFileSystemConfig() {
     Assert.assertEquals(fileSystem.getConf().get(HADOOP_OVERRIDE_PROPERTY_NAME), "value");
   }
 
-  @Test
-  public void testConnectHelixManagerWithRetry() {
+  /**
+   * A helper method that creates a partial instance structure in ZK.
+   */
+  private static void createPartialInstanceStructure(HelixManager helixManager, String zkConnectString) {
     //Connect and disconnect the corrupt task runner to create a Helix Instance set up.
```
Comment (autumnust) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/GobblinTaskRunnerTest.java:148]: This comment no longer accurate: there's no corrupt taskRunner being used here.

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/GobblinTaskRunnerTest.java [hunk 3]
```diff
@@ -156,13 +179,73 @@ public void testConnectHelixManagerWithRetry() {
     //Ensure that connect with retry succeeds
     corruptGobblinTaskRunner.connectHelixManagerWithRetry();
     Assert.assertTrue(true);
+
+    corruptGobblinTaskRunner.disconnectHelixManager();
```
Comment (autumnust) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/GobblinTaskRunnerTest.java:183]: Should `corruptGobblinTaskRunner ` be a local variable?

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/GobblinTaskRunnerTest.java [hunk 4]
```diff
@@ -156,13 +179,73 @@ public void testConnectHelixManagerWithRetry() {
...
+
+    //Ensure that the SleepingTask is running
+    AssertWithBackoff.create().maxSleepMs(100).timeoutMs(2000).backoffFactor(1).
+        assertTrue(ClusterIntegrationTest.isTaskRunning(TaskAssignmentAfterConnectionRetry.TASK_STATE_FILE),"Waiting for the task to enter running state");
+
+    helixManager.disconnect();
   }
 
+
+  public static class TaskAssignmentAfterConnectionRetry extends IntegrationBasicSuite {
```
Comment (autumnust) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/GobblinTaskRunnerTest.java:212]: Can we rename `HelixAssignedParticipantCheckTest#IntegrationJobSuite` to be something generic, and have this class extend that instead ? `overrideJobConfigs` seems to be commonly used when we want to simulate a scenarios when task has to fail somehow.

File: gobblin-cluster/src/test/java/org/apache/gobblin/cluster/GobblinTaskRunnerTest.java [hunk 5]
```diff
@@ -133,21 +141,31 @@ public void testBuildFileSystemConfig() {
     Assert.assertEquals(fileSystem.getConf().get(HADOOP_OVERRIDE_PROPERTY_NAME), "value");
   }
 
-  @Test
-  public void testConnectHelixManagerWithRetry() {
-    //Connect and disconnect the corrupt task runner to create a Helix Instance set up.
+  /**
+   * A helper method that creates a partial instance structure in ZK.
+   */
+  private static void createPartialInstanceStructure(HelixManager helixManager, String zkConnectString) {
```
Comment (autumnust) [inline gobblin-cluster/src/test/java/org/apache/gobblin/cluster/GobblinTaskRunnerTest.java:147]: Should this be moved into utils class newly created ?

**Comment by autumnust** [comment]
+1 LGTM.

**Pull request closed without merging**